modules/sdk-coin-sui/src/lib/transaction.ts

```diff
@@ -254,46 +201,63 @@ export class Transaction extends BaseTransaction {
     }
   }
 
-  private serializeSignedTx(publicKey: BasePublicKey, signature: Buffer): void {
+  /**
+   * Helper function for serialize() to get the correct txData with transaction type
+   */
+  private getTxData(): TxData {
     const suiTx = this._suiTransaction;
-    const txData = {
-      kind: { Single: { Pay: suiTx.payTx } },
+    let tx: TxDetails;
+
+    switch (suiTx.type) {
+      case SuiTransactionType.Pay:
+        tx = {
+          Pay: {
+            coins: suiTx.payTx.coins,
+            recipients: suiTx.payTx.recipients,
+            amounts: suiTx.payTx.amounts,
+          },
+        };
+        break;
+      case SuiTransactionType.PaySui:
+        tx = {
+          PaySui: {
+            coins: suiTx.payTx.coins,
+            recipients: suiTx.payTx.recipients,
+            amounts: suiTx.payTx.amounts,
+          },
+        };
+        break;
+      case SuiTransactionType.PayAllSui:
+        tx = {
+          PayAllSui: {
+            coins: suiTx.payTx.coins,
+            recipient: suiTx.payTx.recipients[0],
+            amount: suiTx.payTx.amounts[0],
+          },
+        };
+        break;
+      default:
+        throw new InvalidTransactionError('SuiTransactionType not supported');
+    }
+
+    return {
+      kind: { Single: tx },
       gasPayment: suiTx.gasPayment,
-      gasPrice: suiTx.gasPrice,
+      gasPrice: SUI_GAS_PRICE,
       gasBudget: suiTx.gasBudget,
       sender: suiTx.sender,
     };
-    const scheme = new Uint8Array([0x00]);
-    const signatureBuffer = Buffer.concat([scheme, signature, Buffer.from(publicKey.pub)]);
-    const signedTx = {
-      data: txData,
-      txSignature: signatureBuffer,
-    };
-
-    const serialized = bcs.ser('SenderSignedData', signedTx).toBytes();
-    const typeTag = Array.from('SenderSignedData::').map((e) => e.charCodeAt(0));
-    const serializedFinal = new Uint8Array(typeTag.length + serialized.length);
-    serializedFinal.set(typeTag);
-    serializedFinal.set(serialized, typeTag.length);
-    this._id = Buffer.from(sha3.sha3_256(serializedFinal), 'hex').toString('base64');
   }
 
   serialize(): string {
-    const suiTx = this._suiTransaction;
-    const txData = {
-      kind: { Single: { Pay: suiTx.payTx } },
-      gasPayment: suiTx.gasPayment,
-      gasPrice: suiTx.gasPrice,
-      gasBudget: suiTx.gasBudget,
-      sender: suiTx.sender,
-    };
+    const txData = this.getTxData();
 
     const dataBytes = bcs.ser('TransactionData', txData).toBytes();
     const serialized = new Uint8Array(TYPE_TAG.length + dataBytes.length);
     serialized.set(TYPE_TAG);
     serialized.set(dataBytes, TYPE_TAG.length);
     if (this._signature !== undefined) {
-      this.serializeSignedTx(this._signature.publicKey, this._signature.signature);
+      this._id = Buffer.from(sha3.sha3_256(serialized), 'hex').toString('base64');
```

Comment:
this looks a bit suspicious. before, we needed the signature to generate the ID, but now we don't?

Reply:
The serialized unsigned and signed tx are the same because of SUI's detached signature. The original `serializeSignedTx` was repeating this `serialize` and setting the same value to `this._id`, so I removed it and set `this._id` here only if the signature exists. If the signature doesn't exist, `this._id` is always UNAVAILABLE.